File: BaekJoon/DFS와 BFS/2667_단지번호붙이기/solve2.py
# dfs

n = int(input())
graph = [list(map(int, input())) for _ in range(n)]
# 이동할 방향 정하기 (상, 하, 좌, 우)
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def dfs(x, y):
    # 공간을 벗어난 경우 return
    if x < 0 or x >= n or y < 0 or y >= n:
        return False
    if graph[x][y] == 1:
        global count
        count += 1
        # 현재 위치 방문 처리
        graph[x][y] = 0
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            dfs(nx, ny)
        return True
    return False

count = 0
result = []

for i in range(n):
    for j in range(n):
        if dfs(i, j):
            result.append(count)
            count = 0
result.sort()
print(len(result))
for res in result:
    print(res)

File: BaekJoon/DFS와 BFS/2667_단지번호붙이기/solve2.java
package com.simplecrud;

import java.io.*;
import java.util.*;

import static java.lang.Integer.*;

public class Main {

    private static int length;
    private static int[][] graph;
    private static boolean[][] visited;
    private static int[] dx = {1, -1, 0, 0};
    private static int[] dy = {0, 0, 1, -1};

    public static void main(String[] args) throws IOException {
        setInput();
        visited = new boolean[length][length];
        final List<Integer> countList = new ArrayList<>();
        for (int i = 0; i < length; i++) {
            for (int j = 0; j < length; j++) {
                if (graph[i][j] == 1 && !visited[i][j]) {
                    countList.add(dfs(i, j, 0));
                }
            }
        }
        Collections.sort(countList);
        System.out.println(countList.size());
        for (Integer count : countList) {
            System.out.println(count);
        }
    }

    private static void setInput() throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        length = parseInt(br.readLine());
        graph = new int[length][length];
        for (int i = 0; i < length; i++) {
            final String[] row = br.readLine().split("");
            for (int j = 0; j < length; j++) {
                graph[i][j] = parseInt(row[j]);
            }
        }
    }

    private static int dfs(int x, int y, int count) {
        visited[x][y] = true;
        count++;
        for (int i = 0; i < 4; i++) {
            final int nx = x + dx[i];
            final int ny = y + dy[i];
            if (0 <= nx && nx < length && 0 <= ny && ny < length) {
                if (graph[nx][ny] == 1 && !visited[nx][ny]) {
                    count = dfs(nx, ny, count);
                }
            }
        }
        return count;
    }
}
